pkg/sql/schemachanger/scdeps/sctestdeps/test_deps.go

```diff
@@ -489,42 +502,109 @@ func (b *testCatalogChangeBatcher) DeleteName(
 
 // ValidateAndRun implements the scexec.CatalogChangeBatcher interface.
 func (b *testCatalogChangeBatcher) ValidateAndRun(ctx context.Context) error {
-	for nameInfo, id := range b.namesToDelete {
+	names := make([]descpb.NameInfo, 0, len(b.namesToDelete))
+	for nameInfo := range b.namesToDelete {
+		names = append(names, nameInfo)
+	}
+	sort.Slice(names, func(i, j int) bool {
+		return b.namesToDelete[names[i]] < b.namesToDelete[names[j]]
+	})
+	for _, nameInfo := range names {
+		expectedID := b.namesToDelete[nameInfo]
 		actualID, hasEntry := b.s.namespace[nameInfo]
 		if !hasEntry {
 			return errors.AssertionFailedf(
 				"cannot delete missing namespace entry %v", nameInfo)
 		}
-		if actualID != id {
+		if actualID != expectedID {
 			return errors.AssertionFailedf(
-				"expected deleted namespace entry %v to have ID %d, instead is %d", nameInfo, id, actualID)
+				"expected deleted namespace entry %v to have ID %d, instead is %d", nameInfo, expectedID, actualID)
 		}
+		b.s.logSideEffectf("delete namespace entry %v -> %d", nameInfo, expectedID)
 		delete(b.s.namespace, nameInfo)
 	}
 	for _, desc := range b.descs {
+		var old catalog.Descriptor
+		if b := descBuilder(b.s.descriptors, desc.GetID()); b != nil {
+			old = b.BuildImmutable()
+		}
+		b.s.logSideEffectf("upsert descriptor #%d%s", desc.GetID(), prettyDescDiff(old, desc, "\n  "))
 		b.s.descriptors.Upsert(desc)
 	}
 	return catalog.Validate(ctx, b.s, catalog.NoValidationTelemetry, catalog.ValidationLevelAllPreTxnCommit, b.descs...).CombinedError()
 }
 
+// prettyDescDiff generates an indented summary of the diff between two
+// descriptors' JSON representations.
+func prettyDescDiff(a, b catalog.Descriptor, indent string) string {
+	toJson := func(d catalog.Descriptor) string {
+		if d == nil {
+			return ""
+		}
+		js, err := protoreflect.MessageToJSON(d.DescriptorProto(), protoreflect.FmtFlags{})
+		if err != nil {
+			panic(err)
+		}
+		str, err := json.Pretty(js)
+		if err != nil {
+			panic(err)
+		}
+		return str
+	}
+
+	d := diff.Diff(toJson(a), toJson(b))
```

Comment:
This is where we use the vendored dependency I'm adding. I couldn't find a better way to diff JSON objects than via their text representation. I think it's fine in practice.